### PR TITLE
FastAPI path operations for Django developers

### PR DESCRIPTION
**Pull request by jerivas**
## Steps to test/reproduce
https://deploy-preview-430--oddleventy.netlify.app/2023/10/24/fastapi-path-operations-for-django-developers/

### CODE REVIEW

**Comment by jerivas** [comment]
[I want a post about FastAPI](https://trello.com/c/1N3xB2el/230-i-want-a-post-about-fastapi)

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+APIs with Python. You might be wondering how it compares to Django, the most
+popular and mature web framework for Python. In this series, I will try to
+answer this question by comparing Django and FastAPI on various aspects and
+features after we successfully converted an internal project from Django to
+FastAPI.
+
+1. FastAPI path operations for Django Developers (this article)
+2. SQL Alchemy for Django Developers (coming soon)
+3. Testing a FastAPI application (coming soon)
+4. How to use FastAPI dependency injection everywhere (coming soon)
```
Comment (jerivas) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:28]: I started writing an article that touched on all these points, but it felt unfocused and too long. I think we have at least three more solid articles to write in this series

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  - FastAPI
+image:
+  src: blog/2023/fast.jpg
+summary: |
+  FastAPI path operations are the equivalent to Django views. In this article we
+  explore the differences, advantages, and gotchas of using them from the
+  perspective of a Django developer.
+---
+
+You might have heard about FastAPI, a modern and fast web framework for building
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:18]: I'd consider linking to FastAPI here where it's first mentioned.

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+APIs with Python. You might be wondering how it compares to Django, the most
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:19]: Repeated use of "you might". Maybe combine these sentences, e.g:

"If you've heard about FastAPI, a modern ... with Python, you might be wondering how it compares..."

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+popular and mature web framework for Python. In this series, I will try to
+answer this question by comparing Django and FastAPI on various aspects and
+features after we successfully converted an internal project from Django to
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:22]:
```suggestion
popular and mature web framework for Python. In this series, I will answer this
question by comparing various aspects and features of Django and FastAPI, based
on our recent experience converting an internal project from Django to FastAPI.
```

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+FastAPI.
+
+1. FastAPI path operations for Django Developers (this article)
+2. SQLAlchemy for Django Developers (coming soon)
+3. Testing a FastAPI application (coming soon)
+4. How to use FastAPI dependency injection everywhere (coming soon)
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:28]:
```suggestion
1. FastAPI Path Operations for Django Developers (this article)
2. SQLAlchemy for Django Developers (coming soon)
3. Testing a FastAPI Application (coming soon)
4. How To Use FastAPI Dependency Injection Everywhere (coming soon)
```

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+## Why is FastAPI worth considering?
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:30]:
```suggestion
## Why Is FastAPI Worth Considering?
```

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+*This section contains opinions. Skip to [Django Views](#django-views) for the
+actual comparison.*
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:33]: I'd probably remove this -- opinions are helpful, and the title of this section already gives the reader a heads-up.

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+applications. First, I expect development tools to do more for me when it comes
+to authoring software. Modern IDEs and code editors have really spoiled me with
+convenient features like go-to-definition, auto-completion, and one-click
+refactoring. This also means I expect languages and frameworks themselves to
+encourage best practices and help me write better code. Static type checking,
+automatic code formatting, and dependency injection are some of the features
+that I have a hard time living without.
+
+Because Django pre-dates Python's type checking system and it (rightly) wants to
+remain as backwards incompatible as possible, all efforts to leverage static
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:53]:
```suggestion
remain as backwards compatible as possible, all efforts to leverage static
```

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+type checking and deeper text editor integration have been bolted-on,
+experimental, and incomplete. The main player in this space seems to be
+[django-stubs], which provides type hints for Django as a separate package.
+After using it for a while, my conclusion is that Django was not designed with
+types in mind, and efforts to add them are mostly futile. We won't go into
+details here, but let's just say that the time and effort of adding and
+maintaining type hints for a Django app is not worth the limited benefits.
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:60]:
```suggestion
types in mind, and efforts to add them are mostly futile. The time and effort of
adding and maintaining type hints for a Django app is not worth the limited
benefits.
```

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+and the need for cohesion and consistency across the API and frontend layers.
+Cohesion means that the API should provide a clear and logical way to access and
+manipulate the data and services that the backend offers. Consistency means that
+the API should follow common standards and conventions for data types, formats,
+errors, validations, and documentation.
+
+Developing APIs with Django means you're probably using the excellent [Django
+REST Framework] (DRF for short). This package is a shining example of how Django
+gives you complete and robust functionality with very little code (shout out to
+you, `ViewSet`). However, we can't deny that it suffers from the same problems
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:74]:
```suggestion
you, `ViewSet`). However, it suffers from the same problems
```

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+as Django itself: it was not designed with types in mind or to share information
+about endpoints and serializers with consumers of its APIs. We tried to bridge
+this gap with [drf-spectacular], which produces [OpenAPI] schemas from DRF views
+and serializers. However, its main limitation is that it relies on developers to
```
Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:78]:
```suggestion
and serializers. Its main limitation is that it relies on developers to
```

**Review by jgerigmeyer — COMMENTED** [review]
This is great! Left a few editing comments. I'm trying to think if there's a way to break up the text with an additional image or two, but that's not a big deal.

**Review by stacyk — APPROVED** [review]
Very informative and easy for someone new to Fast API to understand the differences! 

We have been very consistent with this, but I wonder if we want to start to keep an eye out for image sizing. Its more noticeable on the homepage to see when the aspect ratios are different: 

Screenshot 2023-10-19 at 11 02 03 AM

The style guide states:
The recommended size for hero images is 1600px x 600px. If the aspect ratio for the image is about 3:1 or more, the image will appear shorter on list pages.

I don't think it needs to delay this post at all, but maybe we should add a checklist item or maybe build step/lint step in the future to check this.

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+date: 2023-10-24
+tags:
+  - Article
+  - Python
+  - Django
+  - FastAPI
+image:
+  src: blog/2023/fast.jpg
+summary: |
+  FastAPI path operations are the equivalent to Django views. In this article we
```
Comment (jamesnw) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:13]:
```suggestion
  FastAPI path operations are the equivalent of Django views. In this article we
```

**Review by jamesnw — APPROVED** [review]
Looks great! 

One thought I had is that this talked about how Django was more "batteries included" than Fast API, but as you lay out what FastAPI gives you, I was wondering if it just has a different set of batteries? For instance, you don't need to include `drf-spectacular` anymore, because it is included. However, those might be smaller batteries compared to things like an admin UI.

**Comment by stacyk** [comment]
> We have been very consistent with this, but I wonder if we want to start to keep an eye out for image sizing. Its more noticeable on the homepage to see when the aspect ratios are different:
> 

@jerivas I made a card for making them appear similar but using a fixed height in the grid: https://trello.com/c/WdcjO5hj/234-consider-giving-the-post-grid-article-grid-hero-area-a-fixed-height-so-images-look-consistent-on-post-grid-pages

**Comment by jerivas** [comment]
> I was wondering if it just has a different set of batteries?

@jamesnw that's a great point, I tried to incorporate it in 6c6fdbe1c0d3787317379d4cccd135490682864b

**Review by jgerigmeyer — APPROVED** [review]

**Pull request merged**